ode-wrapper/src/main/java/com/trihydro/odewrapper/helpers/SetItisCodes.java

```diff
@@ -16,352 +15,351 @@
-            if (problemItisCode != null) {
-                items.add(problemItisCode.getItisCode().toString());
-            }
-        }
+    // action
+    IncidentChoice incidentAction = getIncidentActions().stream().filter(x -> x.getCode().equals(wydotTim.getAction())).findFirst().orElse(null);
 
-        if (items.size() == 0)
-            items.add("531");// 531 is "Incident"
-
-        return items;
+    // if action is not null and action itis code exists
+    if (incidentAction != null && incidentAction.getItisCodeId() != null) {
+      getItisCodes().stream().filter(x -> x.getItisCodeId().equals(incidentAction.getItisCodeId())).findFirst()
+          .ifPresent(actionItisCode -> items.add(actionItisCode.getItisCode().toString()));
     }
 
-    public List<IncidentChoice> getIncidentProblems() {
-        if (incidentProblems != null)
-            return incidentProblems;
-        else {
-            incidentProblems = incidentChoicesService.selectAllIncidentProblems();
-            return incidentProblems;
-        }
-    }
+    // effect
+    IncidentChoice incidentEffect = getIncidentEffects().stream().filter(x -> x.getCode().equals(wydotTim.getEffect())).findFirst().orElse(null);
 
-    public List<IncidentChoice> getIncidentEffects() {
-        if (incidentEffects != null)
-            return incidentEffects;
-        else {
-            incidentEffects = incidentChoicesService.selectAllIncidentEffects();
-            return incidentEffects;
-        }
+    // if effect is not null and effect itis code exists
+    if (incidentEffect != null && incidentEffect.getItisCodeId() != null) {
+      getItisCodes().stream().filter(x -> x.getItisCodeId().equals(incidentEffect.getItisCodeId())).findFirst()
+          .ifPresent(effectItisCode -> items.add(effectItisCode.getItisCode().toString()));
     }
 
-    public List<IncidentChoice> getIncidentActions() {
-        if (incidentActions != null)
-            return incidentActions;
-        else {
-            incidentActions = incidentChoicesService.selectAllIncidentActions();
-            return incidentActions;
+    if (wydotTim.getProblem() != null && !wydotTim.getProblem().equals("other")) {
+      // Retrieve the matching incident problem based on the provided code
+      IncidentChoice incidentProblem =
+          getIncidentProblems().stream().filter(problem -> problem.getCode().equals(wydotTim.getProblem())).findFirst().orElse(null);
+
+      // Add the ITIS code if the incident problem exists and has a valid ITIS code ID
+      if (incidentProblem != null) {
+        Integer itisCodeId = incidentProblem.getItisCodeId();
+        if (itisCodeId != null) {
+          getItisCodes().stream().filter(code -> code.getItisCodeId().equals(itisCodeId)).findFirst()
+              .ifPresent(problemItisCode -> items.add(problemItisCode.getItisCode().toString()));
         }
+      }
+    } else {
+      items.addAll(handleOtherIncidentProblem(wydotTim));
     }

```

Comment:
Note: The database only handles one ITIS code per incident right now, so instead of overhauling the system to support multiple codes, I just added the logic for handling "other" incident types directly.